sound/soc/sof/control.c

```diff
@@ -321,25 +321,39 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 		return -EINVAL;
 	}
 
-	if (copy_from_user(cdata->data, tlvd->tlv, header.length))
+	/* first copy only size of sof_abi_hdr */
+	if (copy_from_user(cdata->data, tlvd->tlv, sizeof(struct sof_abi_hdr)))
 		return -EFAULT;
 
-	if (cdata->data->magic != SOF_ABI_MAGIC) {
-		dev_err_ratelimited(scomp->dev,
-				    "error: Wrong ABI magic 0x%08x.\n",
-				    cdata->data->magic);
-		return -EINVAL;
-	}
+	/* check if this blob actually uses sof_abi_hdr */
+	if (cdata->data->magic == SOF_ABI_MAGIC) {
```

Comment:
This doesn't look particularly robust to me, sorry. We check for a magic number, and if we don't find one, we still use the data. I understand that we want to rely on the firmware in such cases to check whether the data is valid. Maybe we should then give the firmware an explicit indication that this data isn't known to the kernel? How about `cdata->data->magic = ~SOF_ABI_MAGIC;`? Or `cdata->data->magic = SOF_ABI_BINARY_LEGACY;` or something?